### PR TITLE
Defer dropdown view initialization until opened and cache shaders between layers

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

Fixes https://www.pivotaltracker.com/story/show/184216698

Reduced impact of node dropdown widgets on load times by deferring creation of grid views until each widget is opened. This also improves node editing time, as the dropdowns are not recreated immediately.

This approach of lazy initialization now caused a significant lag when opening the dropdown. Two major causes of the lag spike is glyph generation (msdfgen, `new_glyph`) and shader compilation (happened every time, because each dropdown has unique layer stack). To reduce the impact of that, the shader compiler now caches the shaders based on generated shader source. Glyph creation hasn't been changed and is still slow. The startup performance is now roughly where it was before introducing widgets.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/gui/view/graph-editor/src/component/node/input/widget.rs [hunk 1]
```diff
@@ -158,12 +166,76 @@ impl SingleChoice {
...
+
+// ====================
+// === LazyDropdown ===
+// ====================
+
+/// A lazy dropdown that is only initialized when it is opened for the first time. This prevents
+/// very long initialization time, as dropdown view creation is currently a very slow process.
+///
+/// Once grid-view creation is reasonably fast, this might be replaced by direct dropdown
+/// initialization on widget creation.
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/node/input/widget.rs]: Comments should explain semantically what code does. This is not a semantic explanation - rather a bug description. It can be left here, but it should be written in a form of a FIXME with a connected issue. Otherwise, it's impossible to track such comments in the code and remove them after the given issue is fixed. Also, sentences such as "is currently a very slow process" are very easy to get obsolete and no one will remember to update the comment.

File: lib/rust/code-builder/src/lib.rs [hunk 2]
```diff
@@ -56,13 +56,17 @@ impl CodeBuilder {
     /// Create a new line and insert appropriate indentation.
     pub fn newline(&mut self) {
         let space_count = self.spaces_in_indent * self.indent;
-        self.write(format!("\n{}", " ".repeat(space_count)));
+        self.buffer.reserve(space_count + 1);
+        self.buffer.push('\n');
+        for _ in 0..space_count {
+            self.buffer.push(' ');
+        }
```
Comment (wdanilo) [inline lib/rust/code-builder/src/lib.rs]: Why this change? The new code is WAY harder to read than the old one.

File: lib/rust/code-builder/src/lib.rs [hunk 3]
```diff
@@ -56,13 +56,17 @@ impl CodeBuilder {
...
         self.spaced = true;
     }
 
     /// Inserts a terminator symbol `;`.
     pub fn terminator(&mut self) {
-        self.write(";");
+        self.buffer.push(';');
```
Comment (wdanilo) [inline lib/rust/code-builder/src/lib.rs]: Why this change? Previous code was using an abstraction `self.write` to update the buffer. This is using internal value. It's longer, more complicated, and less modular. If you want to push a single char, create a function for it, next to `self.write`, or implement a trait which will consume either str or char.

File: lib/rust/code-builder/src/lib.rs [hunk 4]
```diff
@@ -82,12 +86,12 @@ impl CodeBuilder {
...
     /// Specialization of the `add` method for strings.
-    fn add_str<S: Str>(&mut self, s: S) {
+    fn add_str(&mut self, s: &str) {
         if !self.spaced {
-            self.write(" ");
+            self.buffer.push(' ');
         }
         self.spaced = false;
-        self.write(s);
+        self.buffer.push_str(s);
```
Comment (wdanilo) [inline lib/rust/code-builder/src/lib.rs]: Why this change?

File: lib/rust/code-builder/src/lib.rs [hunk 5]
```diff
@@ -82,12 +86,12 @@ impl CodeBuilder {
     }
 
     /// Specialization of the `add` method for strings.
-    fn add_str<S: Str>(&mut self, s: S) {
+    fn add_str(&mut self, s: &str) {
         if !self.spaced {
-            self.write(" ");
+            self.buffer.push(' ');
```
Comment (wdanilo) [inline lib/rust/code-builder/src/lib.rs]: as above

File: lib/rust/code-builder/src/lib.rs [hunk 6]
```diff
@@ -122,7 +126,7 @@ impl AddToBuilder<&str> for CodeBuilder {
 
 impl AddToBuilder<String> for CodeBuilder {
     fn add_to_builder(&mut self, t: String) -> &mut Self {
-        self.add_str(t);
+        self.add_str(&t);
         self
```
Comment (wdanilo) [inline lib/rust/code-builder/src/lib.rs]: if this is changed, then why `add_to_builder` consumes `String`?

File: lib/rust/ensogl/core/src/display/shape/primitive/shader/canvas.rs [hunk 7]
```diff
@@ -37,27 +37,14 @@ impl Shape {
...
 impl ShapeData {
     /// Constructor.
-    pub fn new(shape_num: usize) -> Self {
-        let ids = default();
-        let name = format!("shape_{}", shape_num);
-        Self { shape_num, ids, name }
-    }
-
-    /// Adds new id enclosed in this shape.
-    pub fn add_id(&mut self, id: usize) {
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/shape/primitive/shader/canvas.rs:54]: This is some big change!
1. There is no explanation in the PR why it is done.
2. There is no explanation how it influences the program behavior.
3. I am not sure this is correct now. Anyway, the most important question - why this change was needed and not described?

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 8]
```diff
@@ -82,11 +82,12 @@ const MAX_PARALLEL_COMPILE_JOBS: usize = 2;
...
     #[derivative(Debug = "ignore")]
-    on_ready: Box<dyn FnOnce(shader::Program)>,
+    on_ready:  Box<dyn FnOnce(shader::Program)>,
     #[deref]
-    input:    T,
-    handler:  WeakJobHandler,
-    profiler: profiler::Debug,
+    input:     T,
+    handler:   WeakJobHandler,
+    cache_key: ShaderCacheKey,
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:91]: the name `cache_key` is not explanatory. It needs either better name or documentation.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 9]
```diff
@@ -171,6 +173,7 @@ pub enum Progress {
 pub struct Jobs {
     compile:              Vec<Job<shader::Code>>,
     link:                 Vec<Job<shader::CompiledCode>>,
+    read_cache:           Vec<Job<()>>,
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:178]: `read_cache` sounds like an action, but it is a field. I can't understand what a field with such a name contains.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 10]
```diff
@@ -259,12 +264,36 @@ impl CompilerData {
...
         let handler = strong_handler.downgrade();
         let on_ready = Box::new(on_ready);
-        let job = Job { input, handler, on_ready, profiler };
-        self.jobs.compile.push(job);
+        let cache_key = ShaderCache::code_key(&input);
+
+        // Perform cache lookup before spawning compilation job.
+        match self.cache.get_program(cache_key) {
+            Some(ProgramCacheEntry::Submitted | ProgramCacheEntry::Validated(_)) => {
+                debug!("Reuse cached shader {cache_key:?}.");
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: What does this debug message mean? Does it tell the user to reuse the cached shader? Or does it inform the user that the cached shader will be reused? If the latter, it should be written as "Reusing ...".

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 11]
```diff
@@ -259,12 +264,36 @@ impl CompilerData {
...
+            Some(ProgramCacheEntry::Failed) => {
+                // Shader compilation of this program already failed. Do not submit it again.
+                // This job will never call `on_ready`.
+            }
+            None => {
+                debug!("Compile shader {cache_key:?}.");
+                self.cache.program_submitted(cache_key);
+                let job = Job { input, handler, on_ready, profiler, cache_key };
+                self.jobs.compile.push(job);
+                self.dirty = true;
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs]: why dirty is set here? It should be explained in a comment.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 12]
```diff
@@ -366,13 +396,18 @@ impl CompilerData {
...
             let profiler = job.profiler;
             profiler.resume();
-            let vertex = this.compile_shader(Vertex, job.input.vertex)?;
-            let fragment = this.compile_shader(Fragment, job.input.fragment)?;
+            let vertex = this.cache.get_or_insert_vertex_shader(job.cache_key, || {
+                CompilerData::compile_shader(&this.context, Vertex, job.input.vertex)
+            })?;
+            let fragment = this.cache.get_or_insert_fragment_shader(job.cache_key, || {
+                CompilerData::compile_shader(&this.context, Fragment, job.input.fragment)
+            })?;
```
Comment (wdanilo) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:398]: In both functions you are passing `Vertex` and `Fragment` values. Based on them, the function `get_or_insert_*_shader` could decide whether it is about vex or frag shader, right? Or is it super-complex to be done here?

**Review by wdanilo — CHANGES_REQUESTED** [review]
Regarding the shader cache, I'd love @kazcw to also review it, as he is the author of shader compiler pipeline, so he would probably catch more things than me.

File: lib/rust/code-builder/src/lib.rs (hunk 2 above)
Comment (Frizi) [reply lib/rust/code-builder/src/lib.rs]: The previous implementation allocated two separate temporary strings. The new one has zero temporary allocs and at most once reallocated the string to grow it. This is the one most important change in this file.

I'm going to rewrite it very slightly to use an iterator instead of the loop. I want to keep using `self.buffer`, because it shows the connection with `reserve` above, which is not exposed through `Write` trait.

File: lib/rust/code-builder/src/lib.rs (hunk 3 above)
Comment (Frizi) [reply lib/rust/code-builder/src/lib.rs]: I wanted to avoid the unwraps here. Write/Format traits always return `Result`, but for strings this operation never errors. I'm fine with reducing the number of changes here if you don't like that.

File: lib/rust/code-builder/src/lib.rs (hunk 6 above)
Comment (Frizi) [reply lib/rust/code-builder/src/lib.rs]: This is a trait implementation. That specific impl is on `String` in this case, there are other for refs next to it. In my opinion we shouldn't be implementing this on string, as it potentially hides unnecessary allocations by consuming those strings. But I didn't wanted to change too much outside of this file.

File: lib/rust/ensogl/core/src/display/shape/primitive/shader/canvas.rs (hunk 7 above)
Comment (Frizi) [reply lib/rust/ensogl/core/src/display/shape/primitive/shader/canvas.rs:54]: The whole `ids` vec was simply unused. The only usages were writes or copies from one into another. I simply removed dead code here, also removing some unnecessary vec allocations from shader generator.

About the shape naming change, it used to be based on pointer addresses of temporary values, which were effectively nondeterministic. Each time a shader was built using the same shape definition, the generate code was different due to those names being different. That meant that the shader cache generated different content hash every time. Using a sequential numbering for names fixes that completely. Interestingly, the machinery for that sequential numbering was already in place. The shape deduplication logic still uses the same pointer values for comparison. The only change in shader code output is numeric parts of shape function names.

I'm going to add a comment to the code that assigns the IDs to make it clear what property we expect there.

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 9 above)
Comment (Frizi) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:178]: This follows the naming scheme of other job queues. The job queue is named after an action that the job performs. Is my interpretation of those incorrect?

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs (hunk 12 above)
Comment (Frizi) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:398]: I thought about this, and I think it is not easily possible without implementing a bunch of traits per shader kind on both cache key and cache. That would up being much more code than two separete methods.

Comment (Frizi) [reply lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:398]: This is possible with some extra trait impls on the cache, but it absolutely isn't worth it for two cases. I'm still going to simplify those methods by taking their common part into a `HashMapOps` extension trait, as it is a very generic operation.

File: app/gui/view/graph-editor/src/component/node/input/widget.rs (hunk 1 above)
Comment (Frizi) [reply app/gui/view/graph-editor/src/component/node/input/widget.rs]: Added a fixme with a connected issue.

File: lib/rust/code-builder/src/lib.rs (hunk 6 above)
Comment (Frizi) [reply lib/rust/code-builder/src/lib.rs]: I removed the String implementation of this trait. The only usage of it was for usize, which is better handled by using `write_fmt`, as it doesn't allocate temporary string.

File: lib/rust/code-builder/src/lib.rs (hunk 2 above)
Comment (wdanilo) [reply lib/rust/code-builder/src/lib.rs]: I know, but ugh, its ugly. But yeah, lets keep the performance maxed out.

File: lib/rust/ensogl/core/src/display/shape/primitive/shader/canvas.rs (hunk 7 above)
Comment (wdanilo) [reply lib/rust/ensogl/core/src/display/shape/primitive/shader/canvas.rs:54]: Ok, that makes sense. WOW, I did not realise ids were not used. Hmmm, ok, thanks for clarification. I'd not write this comment if I knew about it.

**Review by wdanilo — APPROVED** [review]

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 13]
```diff
@@ -414,12 +448,39 @@ impl CompilerData {
...
 
+    #[allow(unused_parens)]
+    fn run_next_read_cache_job(&mut self) -> Result<Progress, Error> {
+        let mut progress = Progress::StepProgress;
+        self.with_next_job("shader cache read", (|t| &mut t.jobs.read_cache), |this, job| {
+            job.profiler.resume();
+            // Scheduling cache read job for an entry that was never submitted should not happen.
+            // If it does, this is a bug.
+            let entry = this.cache.get_program(job.cache_key).expect("Shader cache entry missing.");
+            job.profiler.finish();
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:466]: A profiler must not `finish` if it may still be `resume`d. For full accounting, the `resume` above would be paired with a `pause` in the `Submitted` case, and a `finish` in the `Validated`/`Failed` case.

However, I don't think it's worth measuring the time cost of individual hash map lookups. In that case, we don't need to `resume` (or `pause`) in this function--we could just `finish` the profiler in the `Validated` branch (`finish` is legal from the paused state as well as from the resumed state).

File: lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs [hunk 14]
```diff
@@ -429,7 +490,12 @@ impl CompilerData {
...
             if job.handler.exists() {
-                return f(self, job);
+                let cache_key = job.cache_key;
+                let result = f(self, job);
+                if result.is_err() {
+                    self.cache.program_failed(cache_key);
+                }
+                return result;
             } else {
                 trace!("Job handler dropped, skipping.");
```
Comment (kazcw) [inline lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs:500]: I'm concerned about this case. If a `Job` is cancelled (by its handle being dropped), I think it is left in the `Submitted` state in the cache. Will later jobs for the same shader-code encounter that cache entry, which will never be fulfilled, and poll forever?

**Review by kazcw — COMMENTED** [review]
This is a nice optimization! Two bugs noted.

Re: Cancellation, I think handling cancellation might be simpler with a push-based cache solution, e.g. eliminate `Job`'s callback field and store a `Vec<Callback>` in the `ProgramCacheEntry::Submitted` entries, or something along those lines. The other operations use polling queues because the Web API we're working with only supports polling, but we can use a more efficient approach for functionality built on our own APIs. Does that make sense?
